devito/function.py

```diff
@@ -1583,11 +1627,12 @@ def _dist_gather(self, data):
         data = gathered
 
         # Insert back into `self.data` based on the original (expected) data layout
-        self.data[:] = data[self._dist_mask]
+        self.data[:] = data[self._dist_gather_mask]
```

Comment:
`gathered = np.empty(shape=sshape, dtype=self.dtype)`
gahtered is what is received, should be `rshape`